File: 0234-palindrome-linked-list/0234-palindrome-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:

        # 找中間的node(slow)
        slow = head
        fast = head    
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        # 中間之後reverse
        prev = None
        while slow:
            nextNode = slow.next
            slow.next = prev
            prev = slow
            slow = nextNode

        # 比較從最前面(head)到中間和最後面(prev)到中間是否一樣
        while prev: # 用head可能會多一個node
            if head.val != prev.val:
                return False
            head = head.next
            prev = prev.next

        return True

        

File: 0234-palindrome-linked-list/0234-palindrome-linked-list.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

struct ListNode* reverse_list(struct ListNode* head){
    struct ListNode* prev_node = NULL;
    struct ListNode* curr_node = head;
    struct ListNode* next_node = NULL;
    while (curr_node != NULL){
        next_node = curr_node->next;
        curr_node->next = prev_node;
        prev_node = curr_node;
        curr_node = next_node;
    }
    return prev_node;
}

bool isPalindrome(struct ListNode* head) {
    if (head == NULL || head->next == NULL){
        return true;
    }
    struct ListNode* slow = head;
    struct ListNode* fast = head;
    while (fast != NULL && fast->next != NULL){
        slow = slow->next;
        fast = fast->next->next;
    }
    struct ListNode* first_half = head;
    struct ListNode* second_half = reverse_list(slow); // 和first_half一樣長或少一個node
    while (second_half != NULL){
        if (first_half->val != second_half->val){
            return false;
        }
        first_half = first_half->next;
        second_half = second_half->next;
    }
    return true;
}